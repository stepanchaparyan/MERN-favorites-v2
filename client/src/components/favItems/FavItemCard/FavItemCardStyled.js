import styled from 'styled-components';
import ReactTooltip from 'react-tooltip';

export const Container = styled.div`
  display: flex;
  flex-direction: column;
  width: 250px;
  min-width: 250px;
  border: 1px solid ${props => props.theme.cadetblue};
  background-color: WhiteSmoke;
  color: black;
  font-size: 16px;
  margin: 10px;
  &:hover {
    border: 1px solid black;
    background-color: #ebebeb;
  }
`;

export const Info = styled.div`
  display: flex;
  margin-left: 8px;
`;

export const InfoName = styled.div`
  width: 45%;
  padding: 10px 0;
`;

export const InfoData = styled.div`
  padding: 10px 0;
  color: ${props => props.theme.cadetblue};
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  max-width: 120px;
  width: 45%;
`;

export const Button = styled.button`
  border: 0px;
  padding: 4px 0;
  outline: none;
  background-color: ${props => props.theme.cadetblue};
  font-size: 16px;
  margin-top: 5px;
  cursor: pointer;
`;

export const Label = styled.div`
  padding: 8px 0 0 8px;
  width: 8%;
  background-color: ${props => (props.color ? props.color : props.theme.cadetblue)};
  color: white;
  border-bottom-left-radius: 15px;
`;

export const ReactTooltipStyled = styled(ReactTooltip)`
  padding: 4px 6px;
  border-radius: 6px;
`;
