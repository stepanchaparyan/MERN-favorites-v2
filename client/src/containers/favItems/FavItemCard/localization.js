import { defineMessages } from 'react-intl';

export default defineMessages({
  author: {
    id: 'favItem.author',
    defaultMessage: 'Author'
  },
  title: {
    id: 'favItem.title',
    defaultMessage: 'Title'
  },
  category: {
    id: 'favItem.category',
    defaultMessage: 'Category'
  },
  description: {
    id: 'favItem.description',
    defaultMessage: 'Description'
  },
  editCard: {
    id: 'favItem.editCard',
    defaultMessage: 'Edit Card'
  },
  removeCard: {
    id: 'favItem.removeCard',
    defaultMessage: 'Remove Card'
  }
});
