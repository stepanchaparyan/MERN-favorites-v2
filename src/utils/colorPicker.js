import theme from '../styles/theme';
import { FAVITEM } from '../constants';

const { red, blue, limeGreen, orange, black } = theme;
const { TYPE } = FAVITEM;

export const colorPicker = category => {
  let color;
  switch (category) {
    case TYPE.FILM:
      color = red;
      break;
    case TYPE.MUSIC:
      color = blue;
      break;
    case TYPE.BOOKS:
      color = limeGreen;
      break;
    case TYPE.OTHER:
      color = orange;
      break;
    default:
      color = black;
  }
  return color;
};
