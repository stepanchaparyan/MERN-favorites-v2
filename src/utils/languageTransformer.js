const dict = {
  // lang - country
  ar: 'sa',
  be: 'bl',
  cs: 'sz',
  da: 'dk',
  de: 'de',
  el: 'gr',
  en: 'us',
  es: 'es',
  fa: 'ir',
  fr: 'fr',
  ja: 'jp',
  hi: 'in',
  hu: 'hu',
  hr: 'hr',
  hy: 'am',
  it: 'it',
  kk: 'kz',
  nl: 'nl',
  no: 'no',
  pl: 'pl',
  pt: 'pt',
  ru: 'ru',
  sk: 'sk',
  sl: 'sl',
  sv: 'se',
  tr: 'tr',
  uk: 'ua'
};

export const languageTransformer = listList => {
  let newList = [];
  listList.map(lang => {
    const item = Object.keys(dict).find(item => item === lang);
    if (item) {
      newList.push(dict[item].toUpperCase());
    }
  });
  return newList;
};

function getKeyByValue(object, value) {
  return Object.keys(object).find(key => object[key] === value);
}

export const setLanguage = (code, setLocale) => {
  const countryCode = Object.values(dict).includes(code.toLowerCase());

  if (countryCode) {
    setLocale(getKeyByValue(dict, code.toLowerCase()));
  }
};
