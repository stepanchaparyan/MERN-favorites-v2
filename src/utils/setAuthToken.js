import axios from 'axios';

const Authorization = 'Authorization';

const setAuthToken = token => {
  if (token) {
    axios.defaults.headers.common = { Authorization: `Bearer ${token}` };
  } else {
    delete axios.defaults.headers.common[Authorization];
  }
};
export default setAuthToken;
