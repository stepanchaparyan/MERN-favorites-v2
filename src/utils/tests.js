import React from 'react';
import renderer from 'react-test-renderer';
import 'jest-styled-components';
import { shallow } from 'enzyme';

export function renderToJSON(Component, props = {}, Router) {
  if (Router) {
    return renderer
      .create(
        <Router>
          <Component {...props} />
        </Router>
      )
      .toJSON();
  }
  return renderer.create(<Component {...props} />).toJSON();
}

export function renderSnapshot(Component, props = {}) {
  const wrapper = shallow(<Component {...props} />);
  expect(wrapper).toMatchSnapshot();
}
