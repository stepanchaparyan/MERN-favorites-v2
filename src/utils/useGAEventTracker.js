// eslint-disable-next-line no-unused-vars
import React from 'react';
import ReactGA from 'react-ga';

const useGAEventTracker = (category = 'Category') => {
  return (action = 'action', label = 'label') => {
    ReactGA.event({
      category,
      action,
      label,
    });
  };
};

export default useGAEventTracker;
