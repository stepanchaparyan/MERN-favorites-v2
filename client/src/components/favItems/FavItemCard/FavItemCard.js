import React, { useContext } from 'react';
import PropTypes from 'prop-types';
import { useIntl } from 'react-intl';
import { colorPicker } from '../../../utils/colorPicker';
import FavItemContext from '../../../context/favItemContext/favItemContext';
import {
  Container,
  Info,
  InfoName,
  InfoData,
  Button,
  Label,
  ReactTooltipStyled
} from './FavItemCardStyled';
import localization from './localization';
import theme from '../../../styles/theme';
import { REACT_TOOLTIP_STYLED } from '../../../constants';

const FavItemCard = ({ favItem }) => {
  const { cadetblue } = theme;
  const { PLACE, EFFECT } = REACT_TOOLTIP_STYLED;

  const { removeFavItem, edit_FavItem, clearEdit, toggle_Form, toggleForm } = useContext(
    FavItemContext
  );
  const { _id, author, title, category, description } = favItem;
  const { formatMessage } = useIntl();

  const handleRemove = () => {
    removeFavItem(_id);
    clearEdit();
  };

  const handleEdit = () => {
    edit_FavItem(favItem);
    toggle_Form(!toggleForm);
  };

  return (
    <Container>
      <Info>
        <InfoName>{formatMessage(localization.author)}:</InfoName>
        <InfoData data-tip={author} data-arrow-color={cadetblue} data-background-color={cadetblue}>
          {author}
        </InfoData>
        <Label color={colorPicker(category)}>{category.charAt(0)}</Label>
      </Info>
      <Info>
        <InfoName>{formatMessage(localization.title)}:</InfoName>
        <InfoData data-tip={title} data-arrow-color={cadetblue} data-background-color={cadetblue}>
          {title}
        </InfoData>
      </Info>
      <Info>
        <InfoName>{formatMessage(localization.category)}:</InfoName>
        <InfoData>{category}</InfoData>
      </Info>
      <Info>
        <InfoName>{formatMessage(localization.description)}:</InfoName>
        <InfoData
          data-tip={description}
          data-arrow-color={cadetblue}
          data-background-color={cadetblue}
        >
          {description}
        </InfoData>
      </Info>
      <ReactTooltipStyled place={PLACE.LEFT} effect={EFFECT.SOLID} />
      <Button onClick={handleEdit}>{formatMessage(localization.editCard)}</Button>
      <Button onClick={handleRemove}>{formatMessage(localization.removeCard)}</Button>
    </Container>
  );
};

FavItemCard.propTypes = {
  favItem: PropTypes.object.isRequired
};

export default FavItemCard;
