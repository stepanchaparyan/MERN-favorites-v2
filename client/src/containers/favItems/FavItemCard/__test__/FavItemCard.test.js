import React from 'react';
import { renderSnapshot } from '../../../../utils/tests';
import { shallow } from 'enzyme';
import 'jest-styled-components';
import FavItemCard from '../FavItemCard.js';
import { Button } from '../FavItemCardStyled';

jest.mock('react-intl', () => ({
  useIntl: jest.fn(() => ({ formatMessage: m => m.defaultMessage })),
  defineMessages: i => i
}));

jest.mock('react', () => {
  const ActualReact = jest.requireActual('react');
  return {
    ...ActualReact,
    useContext: () => ({
      edit_FavItem: () => {},
      toggle_Form: () => {},
      removeFavItem: () => {},
      clearEdit: () => {}
    })
  };
});

const props = {
  _id: '12345',
  author: 'testAuthor',
  title: 'testTitle',
  category: 'testCategory',
  description: 'testDescription'
};

describe('FavItemCard component test with Enzyme', () => {
  test('renders correct snapshot', () => {
    renderSnapshot(FavItemCard, { favItem: props });
  });

  test('render handleEdit component', () => {
    const component = shallow(<FavItemCard favItem={props} />);
    const button = component.find(Button).at(0);

    button.simulate('click');
    expect(component).toMatchSnapshot();
  });

  test('render handleRemove click', () => {
    const component = shallow(<FavItemCard favItem={props} />);
    const button = component.find(Button).at(1);

    button.simulate('click');
    expect(component).toMatchSnapshot();
  });
});
