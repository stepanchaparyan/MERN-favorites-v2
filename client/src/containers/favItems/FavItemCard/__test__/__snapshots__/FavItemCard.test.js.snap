// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FavItemCard component test with Enzyme render handleEdit component 1`] = `
<styled.div>
  <styled.div>
    <styled.div>
      Author
      :
    </styled.div>
    <styled.div
      data-arrow-color="cadetblue"
      data-background-color="cadetblue"
      data-tip="testAuthor"
    >
      testAuthor
    </styled.div>
    <styled.div
      color="black"
    >
      t
    </styled.div>
  </styled.div>
  <styled.div>
    <styled.div>
      Title
      :
    </styled.div>
    <styled.div
      data-arrow-color="cadetblue"
      data-background-color="cadetblue"
      data-tip="testTitle"
    >
      testTitle
    </styled.div>
  </styled.div>
  <styled.div>
    <styled.div>
      Category
      :
    </styled.div>
    <styled.div>
      testCategory
    </styled.div>
  </styled.div>
  <styled.div>
    <styled.div>
      Description
      :
    </styled.div>
    <styled.div
      data-arrow-color="cadetblue"
      data-background-color="cadetblue"
      data-tip="testDescription"
    >
      testDescription
    </styled.div>
  </styled.div>
  <Styled(ReactTooltip)
    effect="solid"
    place="left"
  />
  <styled.button
    onClick={[Function]}
  >
    Edit Card
  </styled.button>
  <styled.button
    onClick={[Function]}
  >
    Remove Card
  </styled.button>
</styled.div>
`;

exports[`FavItemCard component test with Enzyme render handleRemove click 1`] = `
<styled.div>
  <styled.div>
    <styled.div>
      Author
      :
    </styled.div>
    <styled.div
      data-arrow-color="cadetblue"
      data-background-color="cadetblue"
      data-tip="testAuthor"
    >
      testAuthor
    </styled.div>
    <styled.div
      color="black"
    >
      t
    </styled.div>
  </styled.div>
  <styled.div>
    <styled.div>
      Title
      :
    </styled.div>
    <styled.div
      data-arrow-color="cadetblue"
      data-background-color="cadetblue"
      data-tip="testTitle"
    >
      testTitle
    </styled.div>
  </styled.div>
  <styled.div>
    <styled.div>
      Category
      :
    </styled.div>
    <styled.div>
      testCategory
    </styled.div>
  </styled.div>
  <styled.div>
    <styled.div>
      Description
      :
    </styled.div>
    <styled.div
      data-arrow-color="cadetblue"
      data-background-color="cadetblue"
      data-tip="testDescription"
    >
      testDescription
    </styled.div>
  </styled.div>
  <Styled(ReactTooltip)
    effect="solid"
    place="left"
  />
  <styled.button
    onClick={[Function]}
  >
    Edit Card
  </styled.button>
  <styled.button
    onClick={[Function]}
  >
    Remove Card
  </styled.button>
</styled.div>
`;

exports[`FavItemCard component test with Enzyme renders correct snapshot 1`] = `
<styled.div>
  <styled.div>
    <styled.div>
      Author
      :
    </styled.div>
    <styled.div
      data-arrow-color="cadetblue"
      data-background-color="cadetblue"
      data-tip="testAuthor"
    >
      testAuthor
    </styled.div>
    <styled.div
      color="black"
    >
      t
    </styled.div>
  </styled.div>
  <styled.div>
    <styled.div>
      Title
      :
    </styled.div>
    <styled.div
      data-arrow-color="cadetblue"
      data-background-color="cadetblue"
      data-tip="testTitle"
    >
      testTitle
    </styled.div>
  </styled.div>
  <styled.div>
    <styled.div>
      Category
      :
    </styled.div>
    <styled.div>
      testCategory
    </styled.div>
  </styled.div>
  <styled.div>
    <styled.div>
      Description
      :
    </styled.div>
    <styled.div
      data-arrow-color="cadetblue"
      data-background-color="cadetblue"
      data-tip="testDescription"
    >
      testDescription
    </styled.div>
  </styled.div>
  <Styled(ReactTooltip)
    effect="solid"
    place="left"
  />
  <styled.button
    onClick={[Function]}
  >
    Edit Card
  </styled.button>
  <styled.button
    onClick={[Function]}
  >
    Remove Card
  </styled.button>
</styled.div>
`;
